# Probabilidade e Estatística - Atividade 1

<img src="Arquivos/pe.jpg"  width=80%>

In [171]:
# importar as bibliotecas necessárias
import pandas as pd
import math
from scipy.stats import binom 
from scipy.stats import poisson

## Q1:
Considere uma amostra obtida de uma população e que, nesta amostra, uma das
variáveis era o tempo que o usuário leva para realizar uma determinada tarefa
(variável T). Os tempos e as frequências (número de usuários) são: 10s (7 usuários);
15s (6 usuários); 20s (quatro usuários); 25s (2 usuários) e 30s (apenas um usuário). Por
outro lado, considere a distribuição de probabilidade tal que:

In [172]:
# criar o dataframe
t = [10, 15, 20, 25, 30]
p = [0.35, 0.30, 0.20, 0.10, 0.05]
df = pd.DataFrame(list(zip(t,p)), columns = ['Tempo','Probabilidade'])
df

Tempo  Probabilidade
0     10           0.35
1     15           0.30
2     20           0.20
3     25           0.10
4     30           0.05

Verifique que os resultados da distribuição de frequências para a variável tempo é
equivalente à distribuição de probabilidade acima. Calcule a esperança e a variância da
variável tempo, usando a distribuição de frequências, e constate que E(T) (sua média
ponderada) é igual a 16 e a V(T) = 34.
Explicitar os cálculos para chegar aos valores da esperança e da variância.

### Resposta

### Parte 1: Verifique que os resultados da distribuição de frequências para a variável tempo é equivalente à distribuição de probabilidade acima.

Primeiro irei criar um dataframe com as informações que nos foram dadas sobre a frequência de usuários.

In [173]:
f = [7, 6, 4, 2, 1]
df_new = pd.DataFrame(f, columns = ['Frequência'])
df_new

Frequência
0           7
1           6
2           4
3           2
4           1

In [174]:
total = df_new['Frequência'].sum()
print("A quantidade total de usuários é: ", total)

A quantidade total de usuários é:  20


Para determinar a probabilidade, temos que P = fi/f

In [175]:
probabilidade = []
for fi in df_new['Frequência'].values:
    probabilidade.append(fi/total)

In [176]:
# resultado da probabilidade obtida
probabilidade

[0.35, 0.3, 0.2, 0.1, 0.05]

Por fim, adicionaremos o resultado ao dataframe criado, o que nos dá o seguinte resultado

In [177]:
df_freq = pd.DataFrame(list(zip(df_new['Frequência'].values,probabilidade)), columns = ['Frequência','Probabilidade'])
df_freq

Frequência  Probabilidade
0           7           0.35
1           6           0.30
2           4           0.20
3           2           0.10
4           1           0.05

Concluímos que o valor da probabilidade para a tabela de frequência é equivalente à primeira tabela informada.

### Parte 2: Calcule a esperança e a variância da variável tempo, usando a distribuição de frequências, e constate que E(T) (sua média ponderada) é igual a 16 e a V(T) = 34.

In [178]:
df_esp = pd.DataFrame(list(zip(t, df_new['Frequência'].values)), columns = ['Tempo','Frequência'])
df_esp

Tempo  Frequência
0     10           7
1     15           6
2     20           4
3     25           2
4     30           1

Para encontrarmos a esperança utilizando a tabela de frequência, iremos multiplicar a frequência pelo tempo (xi) em cada linha, e dividir o resultado pela quantidade total de frequência (20).


In [179]:
esperanca = 0
for tempo, freq in zip(df_esp['Tempo'].values, df_esp['Frequência'].values):
    esperanca = tempo * freq + esperanca
esperanca = esperanca / total

In [180]:
print("O valor da esperança é: ", esperanca)

O valor da esperança é:  16.0


In [181]:
# checar o valor de [E(X)]²
valor = esperanca**2
print("O valor de [E(X)]² é ", valor)

O valor de [E(X)]² é  256.0


Para encontrar E(X²), faremos de forma parecida para encontrar a esperança, porém elevando o valor do tempo (xi) ao quadrado

In [182]:
# checar o valor de E(X²)
esperanca2 = 0
for x, y in zip(df_esp['Tempo'].values, df_esp['Frequência'].values):
    esperanca2 = (x**2) * y + esperanca2
esperanca2 = esperanca2 / total
print("O valor de E(X²) é ", esperanca2)

O valor de E(X²) é  290.0


In [183]:
# checar a variância
variancia = esperanca2 - valor
print("A variância se dá por E(X²) - [E(X)]² = ", variancia)

A variância se dá por E(X²) - [E(X)]² =  34.0


Portanto, fica comprovado que E(T) = 16 E V(T) = 34

## Q2
Em Sorocaba, o percentual de pessoas que utilizam regularmente um computador ou um notebook é estimado em cerca de 20%. Uma amostra de 25 pessoas foi retirada desta população e o interesse é saber qual a probabilidade de que menos de três destas pessoas utilizam regularmente um computador ou notebook. E qual a probabilidade de encontrarmos 10 pessoas que utilizam o computador regularmente? E de encontrarmos 18 ou mais pessoas usuárias regulares de computador ou notebook? Com base na amostra, qual o número médio de usuários de computador em Sorocaba (pop.= 700.000)?
Os cálculos podem ser feitos no Excel ou em outro software, mas é necessário explicar os passos que você deu para chegar ao resultado final.

<b> Parte 1:  o interesse é saber qual a probabilidade de que menos de três destas pessoas utilizam regularmente um computador ou notebook. </b>

Utilizaremos para resolver distribuição binomial, onde temos os seguintes dados:
    <ul>
    <li>n = 25</li>
    <li>p = 20% = 0.2 </li>
    <li>k < 3 = (0, 1, 2)</li>
    </ul>

In [184]:
n = 25
p = 0.2
k = [0,1,2]
dist = 0
for r in k:
    dist = binom.pmf(r, n, p) + dist
dist = dist*100
print("A probabilidade de < 3 três pessoas utilizem regularmente um computador é de aproximadamente", round(dist, 2), "%")

A probabilidade de < 3 três pessoas utilizem regularmente um computador é de aproximadamente 9.82 %


<b>Parte 2: E qual a probabilidade de encontrarmos 10 pessoas que utilizam o computador regularmente?</b>

In [185]:
dist = binom.pmf(10, n, p) * 100
print("A probabilidade de encontrarmos 10 pessoas que utilizam o computador regularmente é de aproximadamente", round(dist,2), "%")

A probabilidade de encontrarmos 10 pessoas que utilizam o computador regularmente é de aproximadamente 1.18 %


<b>Parte 3: E de encontrarmos 18 ou mais pessoas usuárias regulares de computador ou notebook?</b>

k >= 18 = 1 - P(k < 18)

In [186]:
dist = 0
for i in range(0,18):
    dist = binom.pmf(i, n, p) + dist
dist = (1 - dist)*100
print("A probabilidade de que 18 ou mais pessoas utilizem regularmente um computador é de aproximadamente", round(dist,10), "%")  


A probabilidade de que 18 ou mais pessoas utilizem regularmente um computador é de aproximadamente 2.9055e-06 %


<b> Parte 4: Com base na amostra, qual o número médio de usuários de computador em Sorocaba? </b>

In [187]:
media = n * p
media_pop = 700.000 * p
print("A média de usuários é", media, "considerando a amostra e é ", media_pop, "considerando a população total.")

A média de usuários é 5.0 considerando a amostra e é  140.0 considerando a população total.


## Q3:
Dados com distribuição Binomial, se temos uma população (N) muito grande, a amostra (n) é grande e o valor de p (probabilidade de sucesso) é pequeno, podem ser aproximados por uma distribuição de Poisson. Na região de Sorocaba, que inclui cidades como Votorantim, Araçoiaba da Serra, etc., com mais de um milhão de habitantes, cerca de 0,01% utiliza um determinado software. Se a distribuição destes usuários é aleatória, qual a probabilidade de encontrarmos cinco usuários deste software em Araçoiaba da Serra, cuja população é igual a 36.000 habitantes? Qual a probabilidade de não haver nenhum usuário deste software em Araçoiaba? E qual a probabilidade de haver mais do que três usuários do referido software em Araçoiaba da Serra? E qual a probabilidade de haver dois ou mais usuários do software em Votorantim, cuja população é de 120000 habitantes?

<b> Parte 1: Se a distribuição destes usuários é aleatória, qual a probabilidade de encontrarmos cinco usuários deste software em Araçoiaba da Serra, cuja população é igual a 36.000 habitantes? </b>

In [188]:
np = 0.01/100 * 36000
k = 5

Temos que:
    <ul>
    <li> np = 3.6 </li>
    <li> k = 5 </li>
    </ul>

In [189]:
dist = poisson.pmf(k, np) * 100
print("A probabilidade é de", dist, "%")

A probabilidade é de 13.768008420609515 %


<b> Parte 2: Qual a probabilidade de não haver nenhum usuário deste software em Araçoiaba? </b>

In [190]:
k = 0
dist = poisson.pmf(k, np) * 100
print("A probabilidade é de", dist, "%")

A probabilidade é de 2.7323722447292558 %


<b> Parte 3: E qual a probabilidade de haver mais do que três usuários do referido software em Araçoiaba da Serra? </b>

k > 3 = 1 - P(k <= 3)

In [191]:
dist = 0
for i in range (0, 4):
    dist = poisson.pmf(i, np) + dist
dist = (1 - dist) * 100
print("A probabilidade é de", dist, "%")

A probabilidade é de 48.478388953385156 %


<b> Parte 4: E qual a probabilidade de haver dois ou mais usuários do software em Votorantim, cuja população é de 120000 habitantes? </b>

k >= 2 = 1 - P(k <2)

In [192]:
np = 0.01/100 * 120000

In [193]:
dist = 0
for i in range(0,2):
    dist = poisson.pmf(i, np) + dist
dist = (1 - dist) * 100
print("A probabilidade é de", dist, "%")

A probabilidade é de 99.99201252394067 %


## Q4 (Distribuição normal):

Considerando que a resistência de uma peça tem distribuição normal com média 24 e variância 49, qual a probabilidade de, ao escolher uma peça ao acaso, tenhamos a resistência:

<ul>
    <li>a) Entre 20 e 24.</li>
    <li>b) Abaixo de 29.</li>
    <li>c) Acima de 33.</li>
    <li>d) Entre 28 e 32.</li>
    <li>e) Qual o intervalo de valores, simétrico em relação à média, que contem aproximadamente 86% das peças?</li>
    <li>f) Qual o intervalo de valores, simétrico em torno da média, que contem 95% das peças.</li>
</ul>


<img src="Arquivos/tabela_dn.png"  width=50%>

Temos então N(24;49)

<b> a) P(20 <= X <= 24) </b>

In [194]:
# P(20 <= X <= 24) = 
Zc =(20 - 24)/math.sqrt(49)
print("ZC = ", round(Zc,2))

ZC =  -0.57


P(-0.57 <= Z <= 0) = P(0 < Z <= 0.57) + P(0) = 0.2157 = 21,57 %

<b>b) P(X <= 29) </b>

In [195]:
Zc = (29 - 24)/math.sqrt(49)
print("ZC = ", round(Zc,2))

ZC =  0.71


P(-infinito <= Z <= 0.71)
<br>
P(-infinito <= Z <= 0) = 0.5
<br>
P(0 <= Z <= 0.71) = 0.2611

P(X <= 29) = 0.5 + 0.2611 = 0.7611 = 76.11 %

<b>c) P(X>=33)</b>

In [196]:
Zc = (33 - 24)/math.sqrt(49)
Zc

1.2857142857142858

P(1.28 <= Z <= infinito)<br>
P(0 <= Z <= infinito) = 0.5<br>
P(0 <= Z <= 1.28) = 0.3997<br>

P(X >=33) = 0.5 - 0.3997 = 0.1003 = 10,03 %

<b> d) P(28 <= X <= 32) </b>

In [197]:
Zc1 =(28 - 24)/math.sqrt(49)
Zc2 =(32 - 24)/math.sqrt(49)
print("Zc1 = ", round(Zc1,2))
print("Zc2 = ", round(Zc2,2))

Zc1 =  0.57
Zc2 =  1.14


P(0.57 <= Z <= 1.14) = P(0 <= Z <= 1.14) + P(0 <= Z <= 0.57)
                     = 0.3729 + 0.2157 = 0.1572 = 58,86 %

<b> e) Qual o intervalo de valores, simétrico em relação à média, que contem aproximadamente 86% das peças </b>

0.4292 -> Zc = 1.47<br>
0.4306 -> Zc = 1.48<br>
0.43 -> ?

(z - 1.47)/(148 - 147) = (0.43 - 0.4292) / (0.4306 - 0.4292)
z = 1.475714

In [198]:
media = 24
sup = 1.475714*math.sqrt(49) + media
inf = -1.475714*math.sqrt(49) + media

print("O intervalo é ", inf, " <= x <= ", sup)

O intervalo é  13.670002  <= x <=  34.329998


<b>f) Qual o intervalo de valores, simétrico em torno da média, que contem 95% das peças?<br>

In [199]:
media = 24
sup = 1.96*math.sqrt(49) + media
inf = -1.96*math.sqrt(49) + media

print("O intervalo é ", inf, " <= x <= ", sup)

O intervalo é  10.280000000000001  <= x <=  37.72


## Q5 Refaça o exercício 3 utilizando a distribuição Normal, considerando duas situações:
<ul>
<li>a. A média μ e a variância σ 2 são obtidas através da distribuição Binomial, ou seja, μ = np e σ 2 = npq. </li>
<li> b. A média e a variância são calculadas a partir da Distribuição de Poisson, ou seja, tanto a média como a variância são iguais a λ = np. </li>
<li> c. Calcule a diferença relativa percentual entre os valores calculados neste e no Exercício3, considerando como base o resultado obtido no Exercício 3.</li>
<li>d. Tanto o resultado obtido no Exercício 3 como estes obtidos nos itens a. e b. deste exercício, são aproximações do resultado exato que seria obtido através da Distribuição Binomial. Verifique se é possível calcular o valor exato, utilizando a Distribuição Binomial e, caso positivo, quais as diferenças relativas entre o resultado exato e os resultados aproximados pela Poisson e pela Normal. </li>
    </ul>

### a. A média μ e a variância σ 2 são obtidas através da distribuição Binomial, ou seja, μ = np e σ 2 = npq.

In [200]:
# LETRA A
media = 0.01/100 * 36000
q = 1 - (0.01/100)
variancia = media * q

print("A média é ", media, " e a variância é ", variancia)

A média é  3.6  e a variância é  3.59964


<b> Parte 3: E qual a probabilidade de haver mais do que três usuários do referido software em Araçoiaba da Serra? </b>

In [201]:
Zc = (3 - media)/math.sqrt(variancia)
print("Zc = ", round(Zc,2))

Zc =  -0.32


P(-0.32 <= Z <= infinito)
P(0 <= Z <= infinito) = 0.5
P(0 <= Z <= 0.32) = 0.1255

P(X >=33) = 0.5 + 0.1255 = 0.6255 (62.55 %)

<b> Parte 4: E qual a probabilidade de haver dois ou mais usuários do software em Votorantim, cuja população é de 120000 habitantes? </b>

In [202]:

media = 0.01/100 * 120000
q = 1 - (0.01/100)
variancia = media * q

Zc = (2 - media)/math.sqrt(variancia)
print("Zc = ", round(Zc,2))

Zc =  -2.89


P(-2.89 <= Z <= infinito)
P(0 <= Z <= infinito) = 0.5
P(0 <= Z <= 2.89) = 0.4981

P(X >=33) = 0.5 + 0.4981 = 0.9981 (99.81 %)

### <li> b. A média e a variância são calculadas a partir da Distribuição de Poisson, ou seja, tanto a média como a variância são iguais a λ = np. </li>

In [203]:
# LETRA B
media = 0.01/100 * 36000
variancia = media 

print("A média é ", media, " e a variância é ", round(variancia,2))

A média é  3.6  e a variância é  3.6


<b> Parte 3: E qual a probabilidade de haver mais do que três usuários do referido software em Araçoiaba da Serra? </b>

In [204]:
Zc = (3 - media)/math.sqrt(variancia)
print("Zc = ", round(Zc,2))

Zc =  -0.32


P(-0.32 <= Z <= infinito)
P(0 <= Z <= infinito) = 0.5
P(0 <= Z <= 0.32) = 0.1255

P(X >=33) = 0.5 + 0.1255 = 0.6255 (62.55 %)

<b> Parte 4: E qual a probabilidade de haver dois ou mais usuários do software em Votorantim, cuja população é de 120000 habitantes? </b>

In [205]:

media = 0.01/100 * 120000
variancia = media 

Zc = (2 - media)/math.sqrt(variancia)
print("Zc = ", round(Zc,2))

Zc =  -2.89


P(-2.89 <= Z <= infinito)
P(0 <= Z <= infinito) = 0.5
P(0 <= Z <= 2.89) = 0.4981

P(X >=33) = 0.5 + 0.4981 = 0.9981 (99.81 %)

### c. Calcule a diferença relativa percentual entre os valores calculados neste e no Exercício3, considerando como base o resultado obtido no Exercício 3.

BASE = B = Resultado da 3
difRel = 100*(resultadoA - resultadoB)/resultadoB


Parte 3: No ex 3 temos 48.478388953385156 % e no 5 temos 62.55 % 

In [206]:
delta = 100*(62.55 - 48.478388953385156)/48.478388953385156
print("A diferença é :", round(delta, 2))

A diferença é : 29.03


Parte 4: No ex 3 temos 99.992  e no 5 temos 99.81

In [207]:
delta = 100*(99.81 - 99.99201252394067)/99.99201252394067
print("A diferença é :", round(delta,4))

A diferença é : -0.182


## d. Tanto o resultado obtido no Exercício 3 como estes obtidos nos itens a. e b. deste exercício, são aproximações do resultado exato que seria obtido através da Distribuição Binomial. Verifique se é possível calcular o valor exato, utilizando a Distribuição Binomial e, caso positivo, quais as diferenças relativas entre o resultado exato e os resultados aproximados pela Poisson e pela Normal.

In [208]:
# Define dados para a binomial
p = 0.01/100 
n = 36000

<b> Parte 1 </b>

In [209]:
dist = binom.pmf(5, n, p) * 100
print("Binomial: ", dist)

Binomial:  13.76858976755183


In [210]:
#Distribuição Relativa Bim e Poisson
bim_poisson = 100*(13.768008420609515 - dist)/dist
print("distRel = ", bim_poisson)

distRel =  -0.004222269325543514


<b> Parte 2 </b>

In [211]:
dist = binom.pmf(0, n, p) * 100
print("Binomial: ", dist)

Binomial:  2.731880429201955


In [212]:
#Distribuição Relativa Bim e Poisson
bim_poisson = 100*(2.7323722447292558 - dist)/dist
print("distRel = ", bim_poisson)

distRel =  0.018002820403243976


<b>Parte 3</b>

In [213]:
dist = 0
for i in range (0, 4):
    dist = binom.pmf(i, n, p) + dist
dist = (1 - dist) * 100
print("Binomial: ", dist)

Binomial:  48.479026393932244


In [214]:
#Distribuição Relativa Bim e Poisson
bim_poisson = 100*(48.478388953385156 - dist)/dist
print("distRel = ", bim_poisson)

distRel =  -0.0013148790198635328


In [215]:
#Distribuição Relativa Bim e Normal
bim_normal = 100 * (62.55 - dist)/dist
print("distRel = ", bim_normal)

distRel =  29.024868386855456


<b>Parte 4:</b>

In [216]:
n = 120000
dist = 0
for i in range(0,2):
    dist = binom.pmf(i, n, p) + dist
dist = (1 - dist) * 100
print("Binomial: ", dist)

Binomial:  99.99201657837149


In [217]:
#Distribuição Relativa Bim e Poisson
bim_poisson = 100*(99.99201252394067 - dist)/dist
bim_poisson
print("distRel = ", bim_poisson)

distRel =  -4.054754525895728e-06


In [218]:
#Distribuição Relativa Bim e Normal
bim_normal = 100 * (99.81 - dist)/dist
print("distRel = ", bim_normal)

distRel =  -0.1820311106825493
